internal/stacksapi/client.go

```diff
@@ -0,0 +1,261 @@
+	DefaultQueue   = "_default"
+)
+
+var (
+	DefaultRetrierOptions = []roko.RetrierOpt{
+		roko.WithMaxAttempts(5),
+		roko.WithStrategy(roko.Constant(1 * time.Second)),
+		roko.WithJitterRange(500*time.Millisecond, 1*time.Second),
+	}
+
+	defaultUserAgent = "agent-stack-k8s/internal/stacksapi/" + version.Version()
+)
+
+func WithLogger(logger *slog.Logger) ClientOpt {
+	return func(c *Client) error {
+		c.logger = logger
+		return nil
+	}
+}
+
+// WithBaseURL sets the base URL for the API client, overriding [DefaultBaseURL]
+func WithBaseURL(baseURL *url.URL) ClientOpt {
+	return func(c *Client) error {
+		c.baseURL = baseURL
+		return nil
+	}
+}
+
+// PrependToUserAgent adds a prefix to the User-Agent header for the API client. Note that overriding of the user-agent
+// is not supported (though can probably be worked around).
+func PrependToUserAgent(prefix string) ClientOpt {
+	return func(c *Client) error {
+		c.userAgent = prefix + " " + c.userAgent
+		return nil
+	}
+}
+
+// WithHTTPClient sets the HTTP client for the API client.
+func WithHTTPClient(httpClient *http.Client) ClientOpt {
+	return func(c *Client) error {
+		c.httpClient = httpClient
+		return nil
+	}
+}
+
+// WithRetrierOptions sets the default retrier options for the API client. These can be overridden per-request by using the [WithRetrier] RequestOpt.
+func WithRetrierOptions(opts ...roko.RetrierOpt) ClientOpt {
+	return func(c *Client) error {
+		c.retrierOpts = opts
+		return nil
+	}
+}
+
+// NewClient creates a new API client with the given API token and options.
+func NewClient(apiToken string, opts ...ClientOpt) (*Client, error) {
+	client := &Client{
+		httpClient:  http.DefaultClient,
+		baseURL:     urlMustParse(DefaultBaseURL),
+		userAgent:   defaultUserAgent,
+		logger:      slog.New(slog.DiscardHandler),
+		authHeader:  fmt.Sprintf("Token %s", apiToken),
+		retrierOpts: DefaultRetrierOptions,
+	}
+
+	for _, opt := range opts {
+		if err := opt(client); err != nil {
+			return nil, err
+		}
+	}
+
+	return client, nil
+}
+
+type StackAPIRequest struct {
+	*http.Request
+
+	retrier   *roko.Retrier
+	bodyBytes []byte
+}
+
+type RequestOption func(*StackAPIRequest) error
+
+func WithRetrier(retrier *roko.Retrier) RequestOption {
+	return func(r *StackAPIRequest) error {
+		r.retrier = retrier
+		return nil
+	}
+}
+
+// NewRequest creates a new API request suitable for dispatch to the Buildkite Stacks API with the given context, method, path, and body.
+func (c *Client) NewRequest(ctx context.Context, method, path string, body any, opts ...RequestOption) (*StackAPIRequest, error) {
+	fullURL := c.baseURL.ResolveReference(&url.URL{Path: path})
+
+	var bodyBytes []byte
+	if body != nil {
+		var err error
+		bodyBytes, err = json.Marshal(body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode request body: %w", err)
+		}
+	}
+
+	req, err := http.NewRequestWithContext(ctx, method, fullURL.String(), bytes.NewReader(bodyBytes))
```

Comment:
it's not technically necessary to set the body here as we re-set it in [`client.Do`](https://github.com/buildkite/agent-stack-k8s/pull/678/files#diff-a900984307a3d164b41fc7e3049c53a6e6f469b7d50bb289e7bbf83ddb64eeb7R186), but IMO it improves the readability of the method and is less surprising than never setting a body. open to other opinions though